### PR TITLE
Fix `non_camel_case_types` for screaming single-words

### LINKED ISSUE #116336: non_camel_case_types does not trigger for screaming single-word
I tried this code:

```rust
#![deny(non_camel_case_types)]

pub struct FOO;

pub enum BAR {
    BAZ,
}

fn main() {}
```

I expected to see this happen: `FOO`, `BAR`, and `BAZ` should all trigger the `non_camel_case_types` lint (at least assuming it follows [RFC430](https://rust-lang.github.io/rfcs/0430-finalizing-naming-conventions.html), and so the code should not compile.

Instead, this happened: The code compiled with no warnings. All three cases start triggering the lint if something like `_X` is added to them.

### Meta

`rustc --version --verbose`:
```
rustc 1.73.0-beta.4 (9f37cd4f7 2023-09-01)
binary: rustc
commit-hash: 9f37cd4f79d289492ccaa8f28733b765498c95e5
commit-date: 2023-09-01
host: aarch64-apple-darwin
release: 1.73.0-beta.4
LLVM version: 17.0.0
```

### PR DESCRIPTION
**Pull request by sjwang05**
Fixes #116336

### CODE REVIEW

File: src/tools/rustfmt/config_proc_macro/src/attrs.rs [hunk 1]
```diff
@@ -68,7 +68,11 @@ fn get_name_value_str_lit(attr: &syn::Attribute, name: &str) -> Option<String> {
     match &attr.meta {
         syn::Meta::NameValue(syn::MetaNameValue {
             path,
-            value: syn::Expr::Lit(syn::ExprLit { lit: syn::Lit::Str(lit_str), .. }),
+            value:
+                syn::Expr::Lit(syn::ExprLit {
+                    lit: syn::Lit::Str(lit_str),
+                    ..
+                }),
```
Comment (ytmimi) [inline src/tools/rustfmt/config_proc_macro/src/attrs.rs:75]: I think it would be better to make this formatting change by opening a PR in https://github.com/rust-lang/rustfmt. When you get a chance please revert this change.

Comment (sjwang05) [reply src/tools/rustfmt/config_proc_macro/src/attrs.rs]: Done, thanks 👍

File: library/std/src/sys/personality/dwarf/eh.rs [hunk 2]
```diff
@@ -43,6 +44,7 @@ pub struct EHContext<'a> {
     pub get_data_start: &'a dyn Fn() -> usize, // Get address of the data section
 }
 
+#[allow(non_camel_case_types)]
 pub enum EHAction {
```
Comment (petrochenkov) [inline library/std/src/sys/personality/dwarf/eh.rs:48]:
```suggestion
pub enum EhAction {
```
`Eh` would be fine in this context.

File: library/std/src/path.rs [hunk 3]
```diff
@@ -177,6 +178,7 @@ pub enum Prefix<'a> {
     /// `\\server\share`.
     ///
     /// UNC prefixes consist of the server's hostname and a share name.
+    #[allow(non_camel_case_types)]
     #[stable(feature = "rust1", since = "1.0.0")]
     UNC(
```
Comment (petrochenkov) [inline library/std/src/path.rs:183]: `Unc` would be fine in this context if these names weren't public and stable.

File: library/core/src/intrinsics.rs [hunk 4]
```diff
@@ -2476,6 +2476,7 @@ extern "rust-intrinsic" {
     /// `unreachable_unchecked` is actually being reached. The bug is in *crate A*,
     /// which violates the principle that a `const fn` must behave the same at
     /// compile-time and at run-time. The unsafe code in crate B is fine.
+    #[allow(non_camel_case_types)]
     #[rustc_const_unstable(feature = "const_eval_select", issue = "none")]
     pub fn const_eval_select<ARG: Tuple, F, G, RET>(
```
Comment (petrochenkov) [inline library/core/src/intrinsics.rs:2481]:
```suggestion
    pub fn const_eval_select<Arg: Tuple, F, G, Ret>(
```
This does look like a clear style violation to me.

File: library/alloc/src/collections/btree/set_val.rs [hunk 5]
```diff
@@ -2,6 +2,7 @@
 /// Used instead of `()` to differentiate between:
 /// * `BTreeMap<T, ()>` (possible user-defined map)
 /// * `BTreeMap<T, SetValZST>` (internal set representation)
+#[allow(non_camel_case_types)]
 #[derive(Debug, Eq, PartialEq, Ord, PartialOrd, Hash, Clone, Default)]
 pub struct SetValZST;
```
Comment (petrochenkov) [inline library/alloc/src/collections/btree/set_val.rs:7]:
```suggestion
pub struct SetValZst;
```
`Zst` is fine in this context.

File: compiler/stable_mir/src/lib.rs [hunk 6]
```diff
@@ -69,6 +69,7 @@ pub type ImplTraitDecls = Vec<ImplDef>;
 #[derive(Clone, Copy, Debug, PartialEq, Eq)]
 pub enum CompilerError<T> {
     /// Internal compiler error (I.e.: Compiler crashed).
+    #[allow(non_camel_case_types)]
     ICE,
```
Comment (petrochenkov) [inline compiler/stable_mir/src/lib.rs:73]:
```suggestion
    Ice,
```
`Ice` is fine in this context.

File: library/core/src/iter/mod.rs [hunk 7]
```diff
@@ -368,6 +368,7 @@ macro_rules! impl_fold_via_try_fold {
         impl_fold_via_try_fold! { @internal spec_rfold -> spec_try_rfold }
     };
     (@internal $fold:ident -> $try_fold:ident) => {
+        #[allow(non_camel_case_types)]
         #[inline]
         fn $fold<AAA, FFF>(mut self, init: AAA, fold: FFF) -> AAA
```
Comment (petrochenkov) [inline library/core/src/iter/mod.rs:373]:
```suggestion
        fn $fold<A, F>(mut self, init: A, fold: F) -> A
```
Not sure why the weird names are used here.

File: compiler/rustc_lint/src/nonstandard_style.rs [hunk 8]
```diff
@@ -93,6 +93,11 @@ fn is_camel_case(name: &str) -> bool {
             // contains a capitalisable character followed by, or preceded by, an underscore
             char_has_case(fst) && snd == '_' || char_has_case(snd) && fst == '_'
         })
+        && !name
+            .chars()
+            .collect::<Vec<_>>()
+            .array_windows()
```
Comment (petrochenkov) [inline compiler/rustc_lint/src/nonstandard_style.rs:99]: Maybe `map_windows` can be used here without collecting?

**Comment by petrochenkov** [comment]
This does look like an improvement to me.
I'm just not sure whether we can make this change at this point though. after treating `ABC` as a standard style for many years.
@rustbot author

File: library/core/src/iter/mod.rs (hunk 7 above)
Comment (sjwang05) [reply library/core/src/iter/mod.rs:375]: Seems like it's because `A` and `F` are already used as generic parameters elsewhere:
```
error[E0403]: the name `A` is already used for a generic parameter in this item's generic parameters
    --> library\core\src\iter\mod.rs:375:18
     |
357  | / macro_rules! impl_fold_via_try_fold {
358  | |     (fold -> try_fold) => {
359  | |         impl_fold_via_try_fold! { @internal fold -> try_fold }
     | |         ------------------------------------------------------ in this macro invocation (#2)
360  | |     };
...    |
375  | |         fn $fold<A, F>(mut self, init: A, fold: F) -> A
     | |                  ^ already used
...    |
383  | |     };
384  | | }
     | | -
     | | |
     | |_in this expansion of `impl_fold_via_try_fold!` (#1)
     |   in this expansion of `impl_fold_via_try_fold!` (#2)
     |
    ::: library\core\src\iter\range.rs:1219:6
     |
1219 |   impl<A: Step> Iterator for ops::RangeInclusive<A> {
     |        - first use of `A`
...
1288 |       impl_fold_via_try_fold! { fold -> try_fold }
     |       -------------------------------------------- in this macro invocation (#1)
```

For now I've kept them as `allow`ed and added a comment explaining that `A` and `F` are already used as generic parameters in impls where this macro appears.

**Comment by sjwang05** [comment]
Thanks for all the feedback!

> I'm just not sure whether we can make this change at this point though. after treating ABC as a standard style for many years.

I can definitely see that--I think I've fixed most of the violations I could find, though it feels like new ones keep popping up whenever I do 😅 . If you (or anyone else) see fit, I'd be more than willing to close this PR because of this.

@rustbot review

**Comment by petrochenkov** [comment]
I'm going to nominate this for the lang team.

A relaxed version of this PR could be bumping the length limit from 3 letters to 4.
E.g. `ABC` would still be considered camel case, but `ABCD`/`ABCDF`/etc not.

**Comment by sjwang05** [comment]
> Can you say more about the motivation here? How essential is this?

I'd say the main motivation here (as mentioned in the original issue) is just to be more compliant with [RFC430](https://rust-lang.github.io/rfcs/0430-finalizing-naming-conventions.html), though I wouldn't say it's essential, especially for 3-letter screaming words.

> Personally, the note from https://github.com/rust-lang/rust/pull/116389#issuecomment-1761569919 about maybe moving the threshold for complaining about it seems potentially interesting as a way to have this handle the most egregious cases without being so noisy.

I agree--I think raising the limit to 4 upper-case characters would be a good compromise, though I'll leave the final call on what to do up to you all.

**Comment by traviscross** [comment]
We discussed this in the T-lang meeting today alongside #60570.  The feeling was that simple heuristics aren't powerful enough to reliably check whether or not some string is following the camel case style since, e.g., a series of upper-case letters may be a series of single letter words or may be an initialism (rather than acronym) that it may be more appropriate to leave capitalized.  As @joshtriplett explains over in #60570:

> I don't think we should _encourage_ `INTEGER_1`, but I _do_ think it's completely valid to write arbitrarily-long strings of capital letters in "camel case". If someone writes `IFAClient` (example made up on the spot), we have no way of knowing if that's "IFA Client" (in which case _officially_ by our rules you're "supposed" to write IfaClient), or "I F A Client" (e.g. Iterator of Futures of Async-wrapped Client). And since we can't distinguish those, we should be more permissive.

And as @scottmcm noted:

> `rb"foo"` being an `RBString` and not getting linted seems entirely fine.

People also felt it might just be too opinionated for `rustc` to enforce `Http` and `Html` over `HTTP` and `HTML`, and that if desired, these may be better as `clippy` lints.  People were also concerned about unnecessary churn here.

With this context, a proposal for FCP close will be forthcoming.

**Comment by tmandry** [comment]
On the basis of the above comment,

@rfcbot fcp close

**Comment by tmandry** [comment]
On the basis of the above comment,

@rfcbot fcp close
